lib/spack/spack/installer.py

```diff
@@ -1411,14 +1474,21 @@ def install(self, **kwargs):
             pkg_id = package_id(pkg)
             tty.verbose('Processing {0}: task={1}'.format(pkg_id, task))
 
+            if task.priority != 0:
+                # Make sure all installed packages are removed from the task's
+                # uninstalled dependencies.  This is necessary at this point
+                # since bootstrap compilers may not be listed as dependents
+                # of packages.
+                task.flag_installed(self.installed)
```

Comment:
If the priority is the number of uninstalled dependents, then I'm concerned there may be queue ordering issues if we have to readjust the priority at this point: ideally after any task is completed, the priority of all related tasks should be updated at that time.

In other words, the priority of tasks in the queue should be updated when we enqueue/complete tasks, not when we retrieve them.

Reply:
Not sure I follow.  

Queue ordering isn't solely dependent on the priority.  `BuildTasks` are ordered by the tuple (`# uninstalled dependencies`, `sequence`), where `sequence` is a counter used to preserve the order in which the task is added (or re-added) to the queue.

Reply:
Just to be clear, I believe the code *is* doing what you think it should in terms of build task priorities.

That is, the priority of the build task is the number of uninstalled dependencies _at the time the task is added or re-queued_ plus a sequence number.  The build task of a dependent package is re-queued when it is determined that one of its dependencies has been "installed".

Reply:
I'll need to read through this code again to refresh my understanding and see if I still have this concern. 

If the task is being requeued when a dependency is installed, I think that would address my concern (and if that is already happening then perhaps I missed it - could you point me to where this happens?). I wasn't sure that was happening though because it didn't look like any requeueing was happening when an installation was finished.